bindings/python/pire.pyx.mako

```diff
@@ -0,0 +1,369 @@
+        return self.scanner.scanner_impl.Final(self.state_impl)
+    % endfor
+
+    def Dead(self):
+        return self.scanner.scanner_impl.Dead(self.state_impl)
+
+    % if "AcceptedRegexps" not in spec.ignored_methods:
+    def AcceptedRegexps(self):
+        cdef:
+            impl.ypair[const size_t*, const size_t*] span
+            impl.yvector[size_t] regexps
+        span = self.scanner.scanner_impl.AcceptedRegexps(self.state_impl)
+        regexps.assign(span.first, span.second)
+        return regexps
+    % endif
+
+    def Step(self, impl.Char ch):
+        check_impl_char(ch)
+        impl.Step(self.scanner.scanner_impl, self.state_impl, ch)
+        return self
+
+    % for method in ["Begin", "End"]:
+    def ${method}(self):
+        impl.Step(self.scanner.scanner_impl, self.state_impl, ${method}Mark)
+        return self
+    % endfor
+
+    def Run(self, bytes line not None):
+        impl.Run(self.scanner.scanner_impl, self.state_impl, begin(line), end(line))
+        return self
+
+    ScannerType = ${Scanner}
+
+    % if Scanner == "CapturingScanner":
+    def Captured(self):
+        if self.state_impl.Captured():
+            return self.state_impl.Begin(), self.state_impl.End()
+        return None
+    % endif
+
+    % if Scanner == "CountingScanner":
+    def Result(self, size_t index):
+        return self.state_impl.Result(index)
+    % endif
+
+
+cdef inline object wrap_${Scanner}(impl.${Scanner} scanner_impl):
+    ret = ${Scanner}()
+    ret.scanner_impl.Swap(scanner_impl)
+    return ret
+
+
+def _${Scanner}_Load(bytes data not None):
+    cdef:
+        impl.yauto_ptr[impl.yistream] stream
+        impl.${Scanner} loaded_scanner
+    stream.reset(new impl.yistream(data))
+    loaded_scanner.Load(stream.get())
+    return wrap_${Scanner}(loaded_scanner)
+
+
+def _reduce_${Scanner}(instance):
+    return _${Scanner}_Load, (instance.Save(),)
+copy_reg.pickle(${Scanner}, _reduce_${Scanner})
+
+
+cdef class ${Scanner}(BaseScanner):
+    cdef impl.${Scanner} scanner_impl
+
+    % if Scanner != "CountingScanner":
+    def __cinit__(self, Fsm fsm=None):
+        if fsm is not None:
+            self.scanner_impl = impl.${Scanner}(fsm.fsm_impl)
+    % else:
+    def __cinit__(self, Fsm pattern=None, Fsm sep=None):
+        if pattern is not None or sep is not None:
+            if pattern is None or sep is None:
+                raise ValueError(
+                    "Expected both pattern and separator, got only one of them"
+                )
+            self.scanner_impl = impl.CountingScanner(pattern.fsm_impl, sep.fsm_impl)
+    % endif
+
+    def Save(self):
+        cdef impl.yostream stream
+        self.scanner_impl.Save(&stream)
+        return stream.GetStr()
+
+    Load = _${Scanner}_Load
+
+    def InitState(self):
+        return ${Scanner}State.__new__(${Scanner}State, self)
+
+    % if "Glue" not in spec.ignored_methods:
+    def GluedWith(self, ${Scanner} rhs not None, size_t max_size=0):
```

Comment:
Without `not None`, the generated code would check the received parameter for something equivalent to `rhs is None or isinstance(rhs, Scanner)`, and would not raise exception for `None` passed here. With `not None` the check is just `isinstance(rhs, Scanner)`.

The typecheck is necessary to early-bind methods and attributes of `rhs`.